crates/wasi-nn/src/backend/winml.rs

```diff
@@ -0,0 +1,165 @@
+//! Implements a `wasi-nn` [`BackendInner`] using WinML.
+
+use super::{BackendError, BackendExecutionContext, BackendFromDir, BackendGraph, BackendInner};
+use crate::wit::types::{ExecutionTarget, GraphEncoding, Tensor};
+use crate::{ExecutionContext, Graph};
+use std::{fs::File, io::Read, path::Path};
+use windows::core::{ComInterface, HSTRING};
+use windows::Storage::Streams::{
+    DataWriter, InMemoryRandomAccessStream, RandomAccessStreamReference,
+};
+use windows::AI::MachineLearning::{
+    LearningModel, LearningModelBinding, LearningModelDevice, LearningModelDeviceKind,
+    LearningModelEvaluationResult, LearningModelSession, TensorFeatureDescriptor, TensorFloat,
+};
+
+#[derive(Default)]
+pub struct WinMLBackend();
+
+impl BackendInner for WinMLBackend {
+    fn encoding(&self) -> GraphEncoding {
+        GraphEncoding::Onnx
+    }
+
+    fn load(&mut self, builders: &[&[u8]], target: ExecutionTarget) -> Result<Graph, BackendError> {
+        if builders.len() != 1 {
+            return Err(BackendError::InvalidNumberOfBuilders(1, builders.len()).into());
+        }
+
+        let model_stream = InMemoryRandomAccessStream::new()?;
+        let model_writer = DataWriter::CreateDataWriter(&model_stream)?;
+        model_writer.WriteBytes(&builders[0])?;
+        model_writer.StoreAsync()?;
+        model_writer.FlushAsync()?;
+        let model = LearningModel::LoadFromStream(&RandomAccessStreamReference::CreateFromStream(
+            &model_stream,
+        )?)?;
+        let device_kind = match target {
+            ExecutionTarget::Cpu => LearningModelDeviceKind::Cpu,
+            ExecutionTarget::Gpu => LearningModelDeviceKind::DirectX,
+            ExecutionTarget::Tpu => unimplemented!(),
+        };
+        let graph = WinMLGraph { model, device_kind };
+
+        let box_: Box<dyn BackendGraph> = Box::new(graph);
+        Ok(box_.into())
+    }
+
+    fn as_dir_loadable(&mut self) -> Option<&mut dyn BackendFromDir> {
+        Some(self)
+    }
+}
+
+impl BackendFromDir for WinMLBackend {
+    fn load_from_dir(
+        &mut self,
+        path: &Path,
+        target: ExecutionTarget,
+    ) -> Result<Graph, BackendError> {
+        let model = read(&path.join("mobilenet.onnx"))?;
```

Comment:
This should handle any kind of model, not just mobilenet. Perhaps just look for the `.onnx` suffix?

Reply:
Renamed to model.onnx to align with openvino backend.